slime/utils/train_infer_is.py

```diff
@@ -0,0 +1,324 @@
+
+        # mode: how to handle the importance sampling weights exceeding the thresholds.
+        if args.train_infer_is_mode == "truncate":
+            # Cap the importance sampling weights at the upper threshold
+            # https://fengyao.notion.site/off-policy-rl#279721e3f6c48092bbe2fcfe0e9c6b33
+            weights = truncate(weights, loss_mask, metrics, args.train_infer_is_upper_bound)
+        elif args.train_infer_is_mode == "mask":
+            # Zero the importance sampling weights outside the [lower, upper] range.
+            # https://yingru.notion.site/When-Speed-Kills-Stability-Demystifying-RL-Collapse-from-the-Training-Inference-Mismatch-271211a558b7808d8b12d403fd15edda
+            weights = mask(
+                weights,
+                loss_mask,
+                metrics,
+                args.train_infer_is_lower_bound,
+                args.train_infer_is_upper_bound,
+            )
+        elif args.train_infer_is_mode == "clip":
+            # Clip the importance sampling weights to the [lower, upper] range.
+            # Original behavior in slime.
+            weights = clip(
+                weights,
+                loss_mask,
+                metrics,
+                args.train_infer_is_lower_bound,
+                args.train_infer_is_upper_bound,
+            )
+        else:
+            raise ValueError(f"Unsupported train_infer_is_mode: {args.train_infer_is_mode}")
+
+        metrics_append(metrics, "ratio_mean_after_tis", weights)
+        if args.train_infer_is_veto_threshold is not None:
+            weights = weights * veto_mask
+            metrics_append(metrics, "ratio_mean_after_veto_mask", weights)
+
+        weights = weights.detach()
+        all_weights.append(weights)
+
+    return all_weights, metrics
+
+
+def compute_train_infer_is_weights_with_cp(
+    args,
+    *,
+    train_log_probs: list[torch.Tensor],
+    rollout_log_probs: list[torch.Tensor],
+    loss_masks: list[torch.Tensor],
+    total_lengths: list[int],
+    response_lengths: list[int],
+) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
+    """
+    Compute the truncated importance sampling (TIS) weights and metrics with context parallel.
+    Args:
+        train_log_probs: List of log probs from training backend on this cp rank. 1D tensor each. Lengths can be different.
+        rollout_log_probs: List of log probs from inference backend on this cp rank. 1D tensor each.
+        loss_masks: List of loss masks. 1D tensor each.
+        total_lengths: List of total lengths.
+        response_lengths: List of response lengths.
+    Returns:
+        is_weights: Importance sampling weights on this CP rank and flattened along dim=0.
+        is_metrics: The metrics for the importance sampling weights, a dict of list[torch.Tensor]. 1D tensor each.
+                    Also flattened along dim=0.
+    """
+    # Gather cp slice from other cp ranks
```

Comment:
I think that from the design philosphy of slime, we won't add much configs for algorithms, but instead, making sure that implementing them are accessible. And also, from the performance point of view, we have go with the zero overhead abstraction idea, which, the user won't pay for the communication or compute overhead unless they require to use the algorithms that have to introduce those. Those are the main ideas we use to make sure slime have a clean and a relatively fast implementation.

With this idea, I think that we can have a default implementation, which was the old and simple token-wise tis without communication. And then add an custom interface to allow users to implement alternative tis methods. (and maybe an example to use cp with such custom interface